reorganize the project

marechal-p opened the pull request:
See commit messages.

marechal-p commented:
cc @PatrickTasse

tahini commented:
Thanks for this, I think there's a path to the base linter file that was not updated though.

marechal-p commented:
@tahini good catch, fixed!

tahini commented on package.json:
```diff
@@ -1,11 +1,17 @@
+    "prepare": "yarn -s tsref && yarn -s tsbuild && yarn -s prepare:packages && yarn -s prepare:examples",
+    "tsref": "node scripts/typescript-references.js",
+    "tsbuild": "tsc -b",
+    "tswatch": "tsc -b -w",
```
What does this compile exactly? Shouldn't it run tswatch/tsbuild in every workspace instead? I tried to make changes in some files, adding a console.log comment in some places. I see that it is recompiling something, but pressing F5 in the browser does not pick up those changes, I still need to compile using the `yarn` command to pick up the changes (but I don't need to rebuild the browser, just pressing F5 after compile works).

marechal-p replied:
> You can also use `yarn tswatch` to watch all TypeScript files and
> rebuild when needed. It only works for TypeScript files, you still need
> to rerun `theia build` to regenerate and build the example applications.

Yes, because this new configuration only takes care of the TypeScript compilation. The bundling step still needs to be done somewhat manually. I think webpack has a watch mode too but I didn't look too much into it.

tahini replied:
ah ok, I can cd examples/browser and run yarn watch and then it works, thanks for the info

tahini reviewed (changes requested):
Debugging from vscode works well for all components, but when debugging directly in the browser (like with F12 in firefox in the debugger tab), the source files are not available as the screenshot shows.

![NoSourceForDebug](https://user-images.githubusercontent.com/4398897/104030893-11c70080-519a-11eb-85cd-b5b86db18a23.png)

marechal-p commented:
@tahini I think the issue with sources in the dev tools is not caused by my changes?

At least I don't see what would have caused it...

tahini reviewed (commented):
Ok, it seems the source files problem was a glitch in the universe, it works fine now.

Could you please add a section in the readme for "Developping the trace extension", saying we should run `yarn tswatch` from the root, then run `yarn watch` in either examples/browser or examples/electron to automatically pick up the changes

tahini reviewed (approved):
I'm fine with this, thanks! I like the watch mode, so that we can only save any file in the repo and simply need to refresh the browser to have it picked up.

But wait for @bhufmann review to see if it changes anything in their internal setup before merging.

marechal-p commented:
Rebased and fixed merge conflicts. Is anything else missing? @PatrickTasse @bhufmann @MatthewKhouzam

marechal-p commented:
After encountering another issue with  @PatrickTasse, we concluded that it is safer to target ES5 when emitting JS.

tahini commented on scripts/typescript-references.js:
```diff
@@ -0,0 +1,72 @@
+const cp = require('child_process')
```
You said on discord that there are some packages now that do this too. Do you wish to keep this file here? Or use one npm package available out there?

marechal-p replied:
I don't see much interest in switching that now. No other new package on npm looks overly popular. Until we have problems I think it's ok to leave this as it is.

bhufmann reviewed (changes requested):
Thanks again. I found some items to fix: 
- Please rebase
- Launch browser back-end/ electron back-end is not working (change program path, outfile path)
- After fixing the back-end launch for me, I was not able to set breakpoints in VsCode in the back-end code. I get "Unbound breakpoint" for each breakpoint. Do we need  sourceMapPathOverrides setting?

marechal-p commented:
@bhufmann thanks for the review, I'll rebase!

> [...] Do we need  sourceMapPathOverrides setting?

I found that it should not be required, I'll look into it after the rebase and make sure everything works.

bhufmann commented:
> @bhufmann thanks for the review, I'll rebase!
> 
> > [...] Do we need  sourceMapPathOverrides setting?
> 
> I found that it should not be required, I'll look into it after the rebase and make sure everything works.

I don't actually know how to fix the unbound breakpoints. I couldn't it make work with sourceMapPathOverrides either. Any ideas?

marechal-p commented:
- Rebased.
- Fixed the debugging issue: it was due to missing `outFiles` entries to glob the `packages` and `theiaExtensions` sources.
- Fixed the plugins not downloading correctly.

bhufmann commented:
Thanks for updating it. The browser back-end debugging works now.

When starting the browser front-end I get an toaster message (see screenshot), which I don't get in master branch. Any ideas?

![image](https://user-images.githubusercontent.com/413424/106616099-26997880-653b-11eb-8df1-de57f9ca807f.png)

marechal-p commented:
@bhufmann it is a weird situation: for some reason old versions of builtin plugins are not available anymore, so I updated the download script to pull the latest version. Issue is, those new versions might try to use unsupported APIs... @marcdumais-work is there something that should be done here?

marcdumais-work commented:
The example app uses @theia `v1.6.0` which is a bit old. Please uplift and then I expect the git extension will work. Or at last not complain at startup :)

marechal-p commented:
Just upgraded dependencies, still need to check the compatibility for some plugins...

marechal-p commented:
@bhufmann it should be mostly fine now. I saw two errors:

1. something about circular references and commands.
2. `languages.registerOnTypeRename` not a function.

So far it looks like non-critical errors, so I would leave as it is.

@vince-fugnitto did I miss anything regarding plugins?

marechal-p commented:
@tahini @bhufmann this PR is re-ready for review.

@vince-fugnitto suggested to refactor the plugins so that they are re-used between both examples, so I did that.

I also removed the incompatible extensions (`html` and `markdown`).

bhufmann commented:
The toaster error message is gone with the latest updates. 
**However I see the following warning when executing yarn:**
`browser-theia-trace-example-app: (Emitted value instead of an instance of Error) Cannot find source file '../../src/protocol/tsp-client.ts': Error: Can't resolve '../../src/protocol/tsp-client.ts' in '/home/eedbhu/git/theia-training/theia-trace-extension/node_modules/tsp-typescript-client/lib/protocol'`

**Running the electron version doesn't work either for me. ** I get the following message when executing "yarn start:electron":
`yarn run v1.21.1
$ yarn rebuild:electron ; yarn --cwd examples/electron start
$ theia rebuild:electron
native node modules are already rebuilt for electron
$ theia start --plugins=local-dir:../plugins
/home/eedbhu/git/theia-training/theia-trace-extension/node_modules/bindings/bindings.js:121
        throw e;
        ^

Error: Module did not self-register: '/home/eedbhu/git/theia-training/theia-trace-extension/node_modules/drivelist/build/Release/drivelist.node'.
    at process.func [as dlopen] (electron/js2c/asar.js:140:31)
    at Object.Module._extensions..node (internal/modules/cjs/loader.js:1034:18)
    at Object.func [as .node] (electron/js2c/asar.js:140:31)
    at Module.load (internal/modules/cjs/loader.js:815:32)
    at Module._load (internal/modules/cjs/loader.js:727:14)
    at Function.Module._load (electron/js2c/asar.js:769:28)
    at Module.require (internal/modules/cjs/loader.js:852:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at bindings (/home/eedbhu/git/theia-training/theia-trace-extension/node_modules/bindings/bindings.js:112:48)
    at Object.<anonymous> (/home/eedbhu/git/theia-training/theia-trace-extension/node_modules/drivelist/js/index.js:25:27) `

marechal-p commented:
> Error: Module did not self-register: '/home/eedbhu/git/theia-training/theia-trace-extension/node_modules/drivelist/build/Release/drivelist.node'.

You need to run `yarn rebuild:electron` before running the electron app?

marechal-p commented:
> browser-theia-trace-example-app: (Emitted value instead of an instance of Error) Cannot find source file '../../src/protocol/tsp-client.ts': Error: Can't resolve '../../src/protocol/tsp-client.ts' in '/home/eedbhu/git/theia-training/theia-trace-extension/node_modules/tsp-typescript-client/lib/protocol'

@bhufmann I looked at the package this warning comes from, and it is indeed missing `.ts` files. Not sure if this is an issue with how the `tsp-typescript-client` is published. Maybe you forgot to include the `src` folder to be published on npm? Fortunately this is just a warning, as long as the `.js` files are published the package is functional :)

I wasn't able to reproduce the Electron issue, could you try to clean your clone and redo?

bhufmann commented:
> > browser-theia-trace-example-app: (Emitted value instead of an instance of Error) Cannot find source file '../../src/protocol/tsp-client.ts': Error: Can't resolve '../../src/protocol/tsp-client.ts' in '/home/eedbhu/git/theia-training/theia-trace-extension/node_modules/tsp-typescript-client/lib/protocol'
> 
> @bhufmann I looked at the package this warning comes from, and it is indeed missing `.ts` files. Not sure if this is an issue with how the `tsp-typescript-client` is published. Maybe you forgot to include the `src` folder to be published on npm? Fortunately this is just a warning, as long as the `.js` files are published the package is functional :)
> 
> I wasn't able to reproduce the Electron issue, could you try to clean your clone and redo?

I did a "git clean -xdf" and "yarn cache clean" before retrying and it worked afterwards.

bhufmann reviewed (changes requested):
I found another thing caused by the new directory structure. When running it in GitPod the browser isn't able to open the workspace. You need to change .gitpod.yml the following line:
`yarn start:browser ../TraceCompassTutorialTraces/ --hostname=0.0.0.0` 
to 
`yarn start:browser ../../TraceCompassTutorialTraces/ --hostname=0.0.0.0`

marechal-p commented:
@bhufmann good catch, fixed the `.gitpod.yml` issue.

bhufmann reviewed (approved):
Looks good to me. Thanks for the contribution.

bhufmann commented:
> > > browser-theia-trace-example-app: (Emitted value instead of an instance of Error) Cannot find source file '../../src/protocol/tsp-client.ts': Error: Can't resolve '../../src/protocol/tsp-client.ts' in '/home/eedbhu/git/theia-training/theia-trace-extension/node_modules/tsp-typescript-client/lib/protocol'
> > 
> > 
> > @bhufmann I looked at the package this warning comes from, and it is indeed missing `.ts` files. Not sure if this is an issue with how the `tsp-typescript-client` is published. Maybe you forgot to include the `src` folder to be published on npm? Fortunately this is just a warning, as long as the `.js` files are published the package is functional :)
> > I wasn't able to reproduce the Electron issue, could you try to clean your clone and redo?
> 
> I did a "git clean -xdf" and "yarn cache clean" before retrying and it worked afterwards.

@marechal-p about the missing source, how to I publish the 'src' folder in the 'tsp-typescript-client' as well? Please let me know.

marcdumais-work commented:
Bernd, I think you need to add the files/folders you need here:
https://github.com/theia-ide/tsp-typescript-client/blob/master/package.json#L27

e.g.:
```json
"files": [
    "lib",
    "src"
  ],
```
